### ISSUE TITLE
Measuring Inference Latency/Throughput

### ISSUE THREAD

**Issue by afzalxo** [pytorch/xla#4405]
## ❓ Questions and Help
Hi,

I am trying to measure the inference latency of a model on TPUv2 and v3. PyTorch.cuda offers a nice interface in the form of `torch.cuda.Event(enable_timing=True)` which allows measuring inference time accurately on GPUs (e.g. [Here](https://deci.ai/blog/measure-inference-time-deep-neural-networks/) is a nice tutorial on this subject which handles GPU sync, allowing accurate measurements).

I'm wondering if there is a similar API that handles timing. I want to measure the time for execution of `output = model(input)` line in the training loop, which would only include the execution time of the TPU for one single inference (i.e., excluding data transfer time between host and device). I would like to obtain this time programmatically, i.e., without tensorboard, since I need to measure latencies for many models in an automated fashion. I can do `time.time` but perhaps it would not be very accurate. Any insights are appreciated.

Thanks a lot!

**Comment by JackCaoG** [comment]
XLA tensor is lazy so this process is a bit special, please take a look at https://github.com/pytorch/xla/blob/master/API_GUIDE.md#xla-tensors-are-lazy first. What you want to do is

```
import torch_xla.core.xla_model as xm

xla_molde = molde.to(xm.xla_device())
# timer start
y = xla_molde(x) # trace the model
xm.mark_step() # execute the model
xm.wait_device_ops() # wait unitl execution finished
# timer end
```

Since we need to trace every time, latency might not be the most ideal. We also recently added experimental support for torch dynamo. https://github.com/pytorch/xla/blob/1858e8bcdd4a3ae1c945763034fef9965ecf3707/test/dynamo/test_dynamo.py#L59-L63

You would need to use nightly to use this feature and this is very experimetnal. Check https://dev-discuss.pytorch.org/t/torchdynamo-update-10-integrating-with-pytorch-xla-for-inference-and-training/935 for more detail.

**Comment by afzalxo** [comment]
Thanks. I realize now that latency is probably not a very useful metric to measure here.

If instead I am measuring the inference throughput, I would like to avoid the warmup period during which the graph is compiled during the initial few iterations of inference (i.e., I would only like to consider the inference throughput for only the later loop iterations for which the compiled program can be fetched from the compilation cache). Would a solution like the following trigger a graph recompilation when the repetition hits the warmup_reps threshold? Or is it good solution to avoid the initial few repetitions when the graph is compiled and stored in the cache? 
```
repetitions = 10   # Inference repetitions to average
warmup_reps = 5    # Warmup inference repetitions to allow caching of compiled graph 
total_time = 0   # Time accumulator to accumulate execution time for each batch
xla_model = model.to(xm.xla_device())
for rep in range(repetitions):
    for data, _ in loader:
       if rep >= warmup_reps:
          start_time = time.time()   # timer start only if warmed up
       y = xla_model(data)
       xm.mark_step()
       xm.wait_device_ops()
       if rep >= warmup_reps:
          total_time += time.time() - start_time   # Accumulate time for this batch
# Calculate throughput using the accumulated time for the repetitions only after the warmup period
throughput = ((repetitions - warmup_reps) * loader.batch_size * len(loader)) / total_time
```

Thanks again!

**Comment by JackCaoG** [comment]
I think this will work.. depending on what you want to measure you can do it differnetly. for example, if you can overlap tracing and execution by doing
```
for _ in warn_up_step:
  y = xla_model(data)

start_time = time.time()
for _ in real_step:
  y = xla_model(data)
  xm.mark_step()
xm.wait_device_ops()
total_time += time.time() - start_time
```

This way during the `real_step` graph execution is overlap with the graph tracing. If you need to have first inference finished before next inference started, you can add `wait_device_ops` in the `for _ in real_step`
